### DISCUSSION TITLE
Allow reading repository secrets from files instead of environment.

### DISCUSSION THREAD

**Discussion by calvinalkan** [semantic-release/semantic-release#2520] [Configuration] [answered]
### New feature motivation

We are using semantic-release in a non-javascript project and everything works great so far.

The only thing I'm not 100% comfortable is passing GitHub credentials as environment variables to semantic-release where any potentially compromised npm dependency of semantic-release could steal them. 

See https://jfrog.com/blog/malicious-npm-packages-are-after-your-discord-tokens-17-new-packages-disclosed/
for recent packages that were taken down for doing precisely this. 

It would be great if we could optionally pass the path to a file on the local filesystem where semantic release would read it from. 

### New feature description

This should not be too complicated to implement. A good idea would be to handle this the way that docker does with secrets where you can either pass `GH_TOKEN` or `GH_TOKEN_PATH` (the latter having priority).

Semantic release could then get the file contents of the passed argument during the config parsing/validation. 

### New feature implementation

At a first glance, this could be implemented in https://github.com/semantic-release/semantic-release/blob/ab45ab1f8d0d0f728fb026a92ec191bcc88f51f1/lib/get-git-auth-url.js#L94

I'm not a JS developer but the process would be: 

1. Loop over all values of the `GIT_TOKENS` const. 
2. For each check if e.g: `GITHUB_TOKEN_PATH` is a readable file, if not check if `GITHUB_TOKEN` is set in the env. 
3. If step 2 succeeded push it to an array called `possibleCredentials`
4. Iterate over all values in `possibleCredentials` and check if the auth URL is valid.
5. Return the first one that is valid (what to do in case of multiple valid ones?) or throw an exception.

**Comment by travi** [comment]
moving the secrets to an unencrypted file does not really make them more secure. sure, it may be a bit more obscure for a malicious package to script in a general manner, but they are still accessible.

instead of security through obscurity, our recommendation is to limit when the secrets are available to your pipeline. for example, this is why our github actions recipe passes the secrets necessary for the release step _only_ to the release step. that way they are not even defined within the pipeline when installing dependencies, which is when postinstall scripts would execute. this is also highlighted as recommended in [this article from the snyk blog](https://snyk.io/blog/github-actions-to-securely-publish-npm-packages/).

**Comment by calvinalkan** [comment, marked as answer]
@travi thanks for the fast response. 

We are already doing all the things you are suggesting.

In our case semantic release runs in docker actually and the GitHub token is provided using docker secrets.

I see your point about security by obscurity but its definately way easier for a compromissed package to just dump the `env` than to scan the entire filesystem for `some/custom/path/to/docker-secrets/GitHub.txt`

**Reply by travi** [reply]
> its definately way easier for a compromissed package to just dump the env than to scan the entire filesystem for `some/custom/path/to/docker-secrets/GitHub.txt`

as i mentioned above, i recognize that type of approach could increase the obscurity. however, that obscurity is mostly present because you are suggesting that it is configurable per project. the suggestion of adding such information to an environment variable like `GITHUB_TOKEN_PATH` reduces that obscurity further. 

part of the problem, however, is that each bit of configurability adds complexity to semantic-release that needs to be maintained over time. we are absolutely supportive of making semantic-release as secure as possible, but what you describe is a bit of an edge case and only minimally reduces risk through obscurity. the ROI of adding such complexity to the project does not seem worth it. even if we were to agree to accepting such a feature, we are a small maintenance team and have other higher priorities, so no timeline could be promised.

again, instead, i am suggesting that it would be more effective to limit the availability of those secrets to the environment as much as you can. since you are using docker, providing the secrets directly as environment variables does not accomplish such a goal since they would be available from the beginning. however, you could potentially source a file that exports them to the environment just before executing the release process.

**Comment by calvinalkan** [comment]
@travi Thanks for the detailed response. It makes sense what you are saying.

"However, you could potentially source a file that exports them to the environment just before executing the release process."

Yea, we are doing something similar by using `docker compose run --env GH_TOKEN=${{ secrets.GH_TOKEN }}`
